# Conexão Drive-Git e Clonagem do Repo no Drive


In [1]:
# 1. Monta o Google Drive (necessário para ter acesso à pasta de destino)
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [2]:
# Dados do Git

github_token = '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'
github_username = 'ArcanjoLucas00'
repo_name = 'Computer_Vision_Cap1'

# Caminho para a pasta no Google Drive que precisa ser atualizada
drive_folder = '/content/drive/MyDrive/computer_vision_cap1'

# 2. Clona a versão mais recente do seu repositório do GitHub
# O comando 'rm -rf' garante que, se a pasta já existir de uma execução anterior,
# ela seja removida para clonar uma cópia limpa e nova.
!rm -rf /content/{repo_name}
!git clone https://{github_username}:{github_token}@github.com/{github_username}/{repo_name}.git

# 3. Sincroniza os arquivos do repositório clonado PARA a pasta do Google Drive
# A origem agora é a pasta do repositório e o destino é a pasta do Drive.
# A opção '--delete' remove arquivos do Drive que não existem mais no repositório,
# tornando a pasta do Drive um espelho exato do seu repo.
print("\nIniciando a sincronização do GitHub para o Google Drive...")
!rsync -av --delete "/content/{repo_name}/" "{drive_folder}/"
print("\nSincronização concluída com sucesso!")

Cloning into 'Computer_Vision_Cap1'...
remote: Enumerating objects: 143, done.
remote: Counting objects: 100% (143/143), done.
remote: Compressing objects: 100% (140/140), done.
remote: Total 143 (delta 4), reused 118 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (143/143), 4.52 MiB | 36.43 MiB/s, done.
Resolving deltas: 100% (4/4), done.

Iniciando a sincronização do GitHub para o Google Drive...
sending incremental file list
./
README.md
.git/
.git/HEAD
.git/config
.git/description
.git/index
.git/packed-refs
.git/branches/
.git/hooks/
.git/hooks/applypatch-msg.sample
.git/hooks/commit-msg.sample
.git/hooks/fsmonitor-watchman.sample
.git/hooks/post-update.sample
.git/hooks/pre-applypatch.sample
.git/hooks/pre-commit.sample
.git/hooks/pre-merge-commit.sample
.git/hooks/pre-push.sample
.git/hooks/pre-rebase.sample
.git/hooks/pre-receive.sample
.git/hooks/prepare-commit-msg.sample
.git/hooks/push-to-checkout.sample
.git/hooks/update.sample
.git/info/
.git/info/exclude
.git/l

In [3]:
# CÉLULA 1: Montar o Google Drive e Instalar Dependências

from google.colab import drive
import os
import json
import shutil
from tqdm import tqdm

# Monta o Google Drive no ambiente do Colab
# Uma janela de autenticação irá aparecer para permitir o acesso.
drive.mount('/content/drive')

# Instala a biblioteca Ultralytics, que contém o YOLOv8.
# O '-q' é para uma instalação silenciosa.
!pip install ultralytics -q

print("Drive montado e Ultralytics instalado com sucesso!")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 26.2 MB/s eta 0:00:00
Drive montado e Ultralytics instalado com sucesso!


# Configuração dos Caminhos e Preparação do Dataset

In [4]:
# CÉLULA 2: Configuração dos Caminhos e Preparação do Dataset (VERSÃO FINAL CORRIGIDA)

import os
import json
import shutil
from tqdm import tqdm

# --- Defina seus caminhos aqui ---
# O caminho para as imagens agora aponta para a subpasta 'train'
gdrive_image_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves/train'

# O caminho para o arquivo JSON permanece na pasta 'chaves'
gdrive_json_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves/labels_my-project-name_2025-10-12-08-37-20.json'

# --- Criação das pastas locais no Colab ---
base_dir = '/content/yolo_dataset'
images_train_dir = os.path.join(base_dir, 'images', 'train')
labels_train_dir = os.path.join(base_dir, 'labels', 'train')

if os.path.exists(base_dir):
    shutil.rmtree(base_dir)

os.makedirs(images_train_dir, exist_ok=True)
os.makedirs(labels_train_dir, exist_ok=True)

print(f"Estrutura de pastas criada em: {base_dir}")

# --- Carregando o arquivo de anotações ---
with open(gdrive_json_path, 'r') as f:
    coco_data = json.load(f)

# --- Processamento das Anotações e Imagens ---
images_info = {img['id']: (img['file_name'], img['width'], img['height']) for img in coco_data['images']}
categories_info = {cat['id']: cat['name'] for cat in coco_data['categories']}
class_names = list(categories_info.values())
num_classes = len(class_names)

print(f"Dataset contém {num_classes} classe(s): {', '.join(class_names)}")

print("\nProcessando anotações e copiando imagens...")
for ann in tqdm(coco_data['annotations']):
    image_id = ann['image_id']
    category_id = ann['category_id']
    bbox = ann['bbox']

    file_name, img_w, img_h = images_info[image_id]

    x_min, y_min, bbox_w, bbox_h = bbox
    x_center = (x_min + bbox_w / 2) / img_w
    y_center = (y_min + bbox_h / 2) / img_h
    norm_w = bbox_w / img_w
    norm_h = bbox_h / img_h

    yolo_class_id = category_id - 1
    yolo_format_str = f"{yolo_class_id} {x_center:.6f} {y_center:.6f} {norm_w:.6f} {norm_h:.6f}\n"

    label_file_name = os.path.splitext(file_name)[0] + '.txt'

    with open(os.path.join(labels_train_dir, label_file_name), 'a') as f:
        f.write(yolo_format_str)

    # **CORREÇÃO APLICADA AQUI**
    # Agora o script busca a imagem na pasta correta ('.../chaves/train/')
    source_image_path = os.path.join(gdrive_image_path, file_name)
    dest_image_path = os.path.join(images_train_dir, file_name)
    if not os.path.exists(dest_image_path):
        shutil.copy(source_image_path, dest_image_path)

print("\nPreparação do dataset concluída!")

Estrutura de pastas criada em: /content/yolo_dataset
Dataset contém 1 classe(s): chave

Processando anotações e copiando imagens...


100%|██████████| 45/45 [00:00<00:00, 204.58it/s]


Preparação do dataset concluída!


##Criando o Arquivo de Configuração do Dataset (YAML)

In [5]:
import yaml

# As variáveis (base_dir, num_classes, class_names) foram definidas na Célula 2
# Este código as utiliza para criar o arquivo de configuração.
yaml_data = {
    'train': os.path.join(base_dir, 'images', 'train'),
    'val': os.path.join(base_dir, 'images', 'train'), # Usando o mesmo conjunto para validação por simplicidade
    'nc': num_classes,                               # Número de classes (neste caso, 1)
    'names': class_names                             # Nomes das classes (['chave'])
}

# Salva o arquivo de configuração no diretório principal do Colab
yaml_file_path = '/content/dataset.yaml'
with open(yaml_file_path, 'w') as f:
    yaml.dump(yaml_data, f, sort_keys=False)

print(f"Arquivo de configuração '{yaml_file_path}' criado com sucesso:")
# O comando 'cat' exibe o conteúdo do arquivo que acabamos de criar
!cat {yaml_file_path}

Arquivo de configuração '/content/dataset.yaml' criado com sucesso:
train: /content/yolo_dataset/images/train
val: /content/yolo_dataset/images/train
nc: 1
names:
- chave


## Inicia o Treinamento do YOLOv8 - CHAVES

In [20]:
from ultralytics import YOLO

# Carrega um modelo pré-treinado.
# 'yolov8n.pt' é o menor e mais rápido. Outras opções: yolov8s, yolov8m, yolov8l, yolov8x
model = YOLO('yolov8n.pt')

# Inicia o treinamento do modelo
# data: Caminho para o arquivo .yaml de configuração.
# epochs: Número de vezes que o modelo verá o dataset inteiro. Comece com 25-50.
# imgsz: Tamanho para o qual as imagens serão redimensionadas para o treinamento. 640 é um bom padrão.
results = model.train(
    data=yaml_file_path,
    epochs=50,
    imgsz=640,
    project='/content/drive/MyDrive/YOLO_Models_Trained', # Nome da pasta onde os resultados serão salvos
    name='chaves_detector'   # Nome da subpasta específica para este treinamento
)

print("\nTreinamento concluído!")
print("Os resultados, incluindo os pesos do modelo treinado, estão na pasta 'yolo_training/chaves_detector'.")

Ultralytics 8.3.214 🚀 Python-3.12.12 torch-2.8.0+cu126 CPU (Intel Xeon CPU @ 2.20GHz)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/dataset_xicaras.yaml, degrees=0.0, deterministic=True, device=cpu, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=chaves_detector2, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, perspective=0.0, plot

# Resumo do Treino - Chaves - 25 Épocas

### mAP50(0.97) - Esta é a métrica principal para detecção de objetos. Um valor de 97% é excelente. Ele mede a precisão do modelo em acertar a localização das "chaves". Simplificando, o modelo está acertando quase sempre.


### mAP50-95(0.866) - Esta é uma métrica mais rigorosa, que mede a precisão em diferentes níveis de exigência de sobreposição. Um valor de 86.6% aqui é muito forte e indica que as caixas delimitadoras (os retângulos) que o modelo desenha estão bem justas aos objetos.

# Resumo do Treino - Chaves - 50 Épocas

### Box(P) - Precisão (0.998): Resultado Praticamente Perfeito. Uma precisão de 99.8% indica que quase todas as detecções que o modelo fez estavam corretas. Ele praticamente não cometeu erros de "falso positivo" (identificar um objeto que não era uma chave). Isso o torna extremamente confiável.

### Box(R) - Recall (1): Resultado Perfeito. Um recall de 100% é a pontuação máxima. Significa que, de todas as chaves que realmente existiam nas 38 imagens, o modelo encontrou todas. Não deixou nenhuma passar despercebida, o que é um feito fantástico.

### mAP50 (0.995): Desempenho de Elite. A métrica principal, que combina precisão e recall, atingiu 99.5%. Este é um resultado de altíssima qualidade, confirmando que o modelo localiza as chaves com uma precisão quase perfeita.

### mAP50-95 (0.977): Excepcional e Robusto. Este é, talvez, o resultado mais impressionante. Um valor de 97.7% nesta métrica, que é muito mais rigorosa, é excepcional. Ele nos diz que as caixas delimitadoras que o modelo desenha não estão apenas no lugar certo, mas também estão muito bem ajustadas ao tamanho e formato das chaves, mesmo sob critérios de avaliação muito exigentes.


# Célula para Validação Dedicada - CHAVES

In [21]:
# CÉLULA DE VALIDAÇÃO FINAL

import os
import json
import shutil
import yaml
from ultralytics import YOLO
from tqdm import tqdm

print("--- Iniciando o Processo de Validação com Dataset Dedicado ---")

# --- 1. CONFIGURAÇÃO DOS CAMINHOS ---
gdrive_val_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves/validation'

# Nome do seu novo arquivo COCO JSON correto
gdrive_val_json_name = 'labels_my-project-name_2025-10-14-02-45-54.json'
gdrive_base_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves'
gdrive_val_json_path = os.path.join(gdrive_base_path, gdrive_val_json_name)

model_path = '/content/yolo_training/chaves_detector/weights/best.pt'

# --- 2. PREPARAÇÃO DO DATASET DE VALIDAÇÃO ---
print(f"\n[PASSO 1/3] Preparando o dataset a partir de: {gdrive_val_path}")

base_dir_val = '/content/yolo_dataset_validation'
images_val_dir = os.path.join(base_dir_val, 'images', 'val')
labels_val_dir = os.path.join(base_dir_val, 'labels', 'val')

if os.path.exists(base_dir_val): shutil.rmtree(base_dir_val)
os.makedirs(images_val_dir, exist_ok=True)
os.makedirs(labels_val_dir, exist_ok=True)

try:
    print(f"Carregando anotações de: {gdrive_val_json_path}")
    with open(gdrive_val_json_path, 'r') as f:
        coco_data = json.load(f)

    if 'annotations' not in coco_data or 'images' not in coco_data or 'categories' not in coco_data:
        raise ValueError("Erro: O arquivo JSON não parece estar no formato COCO.")

    images_info = {img['id']: (img['file_name'], img['width'], img['height']) for img in coco_data['images']}
    categories_info = {cat['id']: cat['name'] for cat in coco_data['categories']}
    class_names = list(categories_info.values())
    num_classes = len(class_names)

    print(f"Encontradas {len(coco_data['images'])} imagens e {len(coco_data['annotations'])} anotações no JSON.")

    valid_annotations_found = 0
    for ann in tqdm(coco_data['annotations'], desc="Processando anotações"):
        if 'category_id' not in ann or 'bbox' not in ann:
            continue

        valid_annotations_found += 1
        image_id = ann['image_id']
        file_name, img_w, img_h = images_info[image_id]

        source_image_path = os.path.join(gdrive_val_path, file_name)
        dest_image_path = os.path.join(images_val_dir, file_name)
        if not os.path.exists(dest_image_path):
            try:
                shutil.copy(source_image_path, dest_image_path)
            except FileNotFoundError:
                print(f"\nAviso: A imagem '{file_name}' não foi encontrada em '{gdrive_val_path}' e será ignorada.")
                continue

        label_file_name = os.path.splitext(file_name)[0] + '.txt'
        with open(os.path.join(labels_val_dir, label_file_name), 'a') as f:
            category_id = ann['category_id']
            bbox = ann['bbox']
            x_min, y_min, bbox_w, bbox_h = bbox
            x_center = (x_min + bbox_w / 2) / img_w
            y_center = (y_min + bbox_h / 2) / img_h
            norm_w = bbox_w / img_w
            norm_h = bbox_h / img_h
            yolo_class_id = category_id - 1
            yolo_format_str = f"{yolo_class_id} {x_center:.6f} {y_center:.6f} {norm_w:.6f} {norm_h:.6f}\n"
            f.write(yolo_format_str)

    if valid_annotations_found == 0:
        raise ValueError("Nenhuma anotação válida foi encontrada. Por favor, re-exporte no formato COCO.")

    print("Dataset de validação preparado com sucesso.")

    # --- 3. CRIAÇÃO DO ARQUIVO YAML ---
    print("\n[PASSO 2/3] Criando arquivo de configuração 'validation.yaml'")
    yaml_val_data = {
        'train': images_val_dir,
        'val': images_val_dir,
        'nc': num_classes,
        'names': class_names
    }
    yaml_val_file_path = '/content/validation.yaml'
    with open(yaml_val_file_path, 'w') as f: yaml.dump(yaml_val_data, f, sort_keys=False)
    !cat {yaml_val_file_path}

    # --- 4. EXECUÇÃO DA VALIDAÇÃO ---
    print("\n[PASSO 3/3] Executando a validação...")
    model = YOLO(model_path)
    validation_results = model.val(data=yaml_val_file_path)
    print("\n--- Processo de Validação Concluído ---")

except (FileNotFoundError, ValueError) as e:
    print(f"\nERRO CRÍTICO: {e}")

--- Iniciando o Processo de Validação com Dataset Dedicado ---

[PASSO 1/3] Preparando o dataset a partir de: /content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves/validation
Carregando anotações de: /content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves/labels_my-project-name_2025-10-14-02-45-54.json
Encontradas 6 imagens e 6 anotações no JSON.


Processando anotações: 100%|██████████| 6/6 [00:00<00:00, 169.99it/s]

Dataset de validação preparado com sucesso.

[PASSO 2/3] Criando arquivo de configuração 'validation.yaml'
train: /content/yolo_dataset_validation/images/val
val: /content/yolo_dataset_validation/images/val
nc: 1
names:
- chave

[PASSO 3/3] Executando a validação...


Ultralytics 8.3.214 🚀 Python-3.12.12 torch-2.8.0+cu126 CPU (Intel Xeon CPU @ 2.20GHz)
Model summary (fused): 72 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 681.7±196.5 MB/s, size: 50.7 KB)
val: Scanning /content/yolo_dataset_validation/labels/val... 6 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 6/6 938.0it/s 0.0s
val: New cache created: /content/yolo_dataset_validation/labels/val.cache
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 1/1 0.5it/s 1.9s
                   all          6          6       0.79      0.635      0.533      0.112
Speed: 2.3ms preprocess, 284.7ms inference, 0.0ms loss, 16.5ms postprocess per image
Results saved to /content/runs/detect/val5

--- Processo de Validação Concluído ---


# Resumo da Validação Dedicada - 25 Épocas:

### Box(P) - precisão (0.79): Das vezes que o modelo disse "isto é uma chave", ele estava certo em 79% das vezes. As outras 21% foram "falsos positivos" (ele detectou algo que não era uma chave, ou detectou a mesma chave mais de uma vez). Este é um bom número.

### Box(R) - Recall(0.635): De todas as chaves que realmente existiam nas 6 imagens, o modelo conseguiu encontrar 63.5% delas. Os 36.5% restantes são "falsos negativos" (chaves que estavam na imagem, mas que o modelo não detectou).

### mAP50(0.533):  Esta é a métrica principal. Ela representa a "Média da Precisão Média" com um limiar de acerto de 50%. Simplificando, é uma nota geral que equilibra a precisão e o recall. Um resultado de 53.3% é um ponto de partida razoável. Ele nos diz que o modelo aprendeu o conceito de "chave", mas ainda não é um especialista e comete erros ao ser exposto a imagens novas.

### mAP50-95(0.112): Esta é uma métrica muito mais rigorosa, que calcula a média do mAP em diferentes níveis de exigência (de 50% até 95%). O valor baixo de 11.2% indica que, embora o modelo consiga encontrar a chave, as caixas delimitadoras (os retângulos) que ele desenha muitas vezes não estão perfeitamente alinhadas com o objeto real.

# Resumo da Validação Dedicada - 50 Épocas:

### Box(P) - Precisão (0.79): De todas as vezes que o modelo disse "isto é uma chave", ele estava certo em 79% das vezes. Este é um bom resultado. Significa que o modelo tem uma alta confiança e raramente comete erros de "falso positivo" (identificar algo que não é uma chave).

### Box(R) - Recall (0.635):De todas as chaves que realmente existiam nas imagens, o modelo conseguiu encontrar 63.5% delas.Este resultado é moderado. Significa que o modelo deixou de detectar cerca de 36.5% das chaves que estavam presentes (falsos negativos). Ele é bom em acertar quando detecta, mas não é tão bom em encontrar todos os objetos.

### mAP50 (0.533): Esta é a métrica principal (Média da Precisão Média) é a nota geral que equilibra a precisão e o recall. O "50" significa que uma detecção é considerada "correta" se a caixa desenhada pelo modelo tiver pelo menos 50% de sobreposição com a caixa real. Uma pontuação de 53.3% indica que o modelo é um protótipo funcional. Ele claramente aprendeu o que é uma chave, mas ainda não é um especialista. Ele comete erros e tem espaço significativo para melhorar.

### mAP50-95 (0.112):Esta é uma métrica muito mais rigorosa. Ela calcula a média do mAP em 10 níveis diferentes de exigência (de 50% a 95% de sobreposição). O valor baixo de 11.2% nos diz que as caixas delimitadoras que o modelo desenha não são muito precisas. Elas são boas o suficiente para passar no teste de 50% (por isso o mAP50 é 0.533), mas falham nos testes mais rigorosos.

# Teste: Modelo do Conjunto de Dados de Teste - CHAVES

In [22]:
# CÉLULA DE TESTE FINAL (com o caminho do modelo corrigido)

import os
import json
import shutil
import yaml
from ultralytics import YOLO
from tqdm import tqdm

print("--- Iniciando o Processo de Teste Final do Modelo ---")

# --- 1. CONFIGURAÇÃO DOS CAMINHOS ---
gdrive_test_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves/test'
gdrive_base_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves'
gdrive_test_json_name = 'labels_my-project-name_2025-10-14-03-27-51.json'
gdrive_test_json_path = os.path.join(gdrive_base_path, gdrive_test_json_name)

# **CORREÇÃO APLICADA AQUI: Apontando para a pasta correta no Drive**
model_path = '/content/drive/MyDrive/YOLO_Models_Trained/chaves_detector/weights/best.pt'

# --- 2. PREPARAÇÃO DO DATASET DE TESTE ---
print(f"\n[PASSO 1/3] Preparando o dataset de teste a partir de: {gdrive_test_path}")

base_dir_test = '/content/yolo_dataset_test'
images_test_dir = os.path.join(base_dir_test, 'images', 'test')
labels_test_dir = os.path.join(base_dir_test, 'labels', 'test')

if os.path.exists(base_dir_test): shutil.rmtree(base_dir_test)
os.makedirs(images_test_dir, exist_ok=True)
os.makedirs(labels_test_dir, exist_ok=True)

try:
    print(f"Carregando anotações de: {gdrive_test_json_path}")
    with open(gdrive_test_json_path, 'r') as f:
        coco_data = json.load(f)

    if 'annotations' not in coco_data or 'images' not in coco_data or 'categories' not in coco_data:
        raise ValueError("Erro: O arquivo JSON não parece estar no formato COCO.")

    images_info = {img['id']: (img['file_name'], img['width'], img['height']) for img in coco_data['images']}
    categories_info = {cat['id']: cat['name'] for cat in coco_data['categories']}
    class_names = list(categories_info.values())
    num_classes = len(class_names)

    print(f"Encontradas {len(coco_data['images'])} imagens e {len(coco_data['annotations'])} anotações no JSON de teste.")

    for ann in tqdm(coco_data['annotations'], desc="Processando anotações de teste"):
        if 'category_id' not in ann or 'bbox' not in ann: continue
        image_id = ann['image_id']
        file_name, img_w, img_h = images_info[image_id]
        source_image_path = os.path.join(gdrive_test_path, file_name)
        dest_image_path = os.path.join(images_test_dir, file_name)
        if not os.path.exists(dest_image_path):
            try: shutil.copy(source_image_path, dest_image_path)
            except FileNotFoundError: continue
        label_file_name = os.path.splitext(file_name)[0] + '.txt'
        with open(os.path.join(labels_test_dir, label_file_name), 'a') as f:
            category_id = ann['category_id']; bbox = ann['bbox']
            x_min, y_min, bbox_w, bbox_h = bbox
            x_center = (x_min + bbox_w / 2) / img_w; y_center = (y_min + bbox_h / 2) / img_h
            norm_w = bbox_w / img_w; norm_h = bbox_h / img_h
            yolo_class_id = category_id - 1
            yolo_format_str = f"{yolo_class_id} {x_center:.6f} {y_center:.6f} {norm_w:.6f} {norm_h:.6f}\n"
            f.write(yolo_format_str)
    print("Dataset de teste preparado com sucesso.")

    # --- 3. CRIAÇÃO DO ARQUIVO YAML ---
    print("\n[PASSO 2/3] Criando arquivo de configuração 'test.yaml'")
    yaml_test_data = {'train': images_test_dir, 'val': images_test_dir, 'nc': num_classes, 'names': class_names}
    yaml_test_file_path = '/content/test.yaml'
    with open(yaml_test_file_path, 'w') as f: yaml.dump(yaml_test_data, f, sort_keys=False)
    !cat {yaml_test_file_path}

    # --- 4. EXECUÇÃO DO TESTE FINAL ---
    print("\n[PASSO 3/3] Executando o teste final com o modelo treinado...")
    model = YOLO(model_path)
    test_results = model.val(data=yaml_test_file_path)
    print("\n--- Processo de Teste Concluído ---")

except (FileNotFoundError, ValueError) as e:
    print(f"\nERRO CRÍTICO: {e}")

--- Iniciando o Processo de Teste Final do Modelo ---

[PASSO 1/3] Preparando o dataset de teste a partir de: /content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves/test
Carregando anotações de: /content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/chaves/labels_my-project-name_2025-10-14-03-27-51.json
Encontradas 6 imagens e 6 anotações no JSON de teste.


Processando anotações de teste: 100%|██████████| 6/6 [00:00<00:00, 173.19it/s]

Dataset de teste preparado com sucesso.

[PASSO 2/3] Criando arquivo de configuração 'test.yaml'
train: /content/yolo_dataset_test/images/test
val: /content/yolo_dataset_test/images/test
nc: 1
names:
- chave

[PASSO 3/3] Executando o teste final com o modelo treinado...


Ultralytics 8.3.214 🚀 Python-3.12.12 torch-2.8.0+cu126 CPU (Intel Xeon CPU @ 2.20GHz)
Model summary (fused): 72 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 892.8±141.6 MB/s, size: 56.4 KB)
val: Scanning /content/yolo_dataset_test/labels/test... 6 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 6/6 667.0it/s 0.0s
val: New cache created: /content/yolo_dataset_test/labels/test.cache
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 1/1 0.7it/s 1.4s
                   all          6          6      0.564      0.441      0.534      0.157
Speed: 3.2ms preprocess, 208.4ms inference, 0.0ms loss, 15.2ms postprocess per image
Results saved to /content/runs/detect/val6

--- Processo de Teste Concluído ---


# Resumo do Teste Final - 25 Épocas


### Box(P) - Precisão (0.564): A precisão de 56.4% é um resultado modesto. Significa que, quando o modelo detectou uma "chave" nas imagens de teste, ele estava correto em pouco mais da metade das vezes. Os outros ~44% foram detecções incorretas (falsos positivos).

### Box(R) - Recall (0.441): O recall de 44.1% indica que o modelo conseguiu encontrar menos da metade de todas as chaves que realmente existiam nas imagens de teste. Ele deixou passar um número significativo de objetos (falsos negativos).

### mAP50 (0.534): Esta é a sua métrica principal e a "nota final" do modelo. Um resultado de 53.4% mostra que o modelo tem uma capacidade básica de identificar chaves, mas seu desempenho cai consideravelmente quando exposto a dados completamente novos, indicando que ele não é muito robusto.

### mAP50-95 (0.157): A baixa pontuação de 15.7% nesta métrica mais rigorosa confirma que as detecções que o modelo acertou não foram muito precisas em termos de localização (as caixas delimitadoras não estavam bem ajustadas).


### Considerações: Fazendo uma análise das imagens utilizadas em todas as etapas, foi constatado que as imagens utilizadas para o teste, possuiam um modelo de chave pouco visto pelo modelo nas etapas anteriores. Apesar de ainda haver reconhecimento, pode ter influenciado para um reconhecimento mais dificultoso, uma vez que não houveram tantos modelos parecidos para treino e validação.

### A conclusão mais importante aqui é a grande diferença entre o desempenho do modelo nos dados de treino (~97%) e nos dados de teste (53%). Este fenômeno é chamado de overfitting (sobreadaptação).Significa que modelo "decorou" muito bem as imagens de treino, mas não aprendeu a "generalizar" o conceito de "chave" de forma eficaz. Quando ele vê imagens novas e um pouco diferentes no conjunto de teste, ele tem dificuldade em aplicar o que aprendeu.

# Conclusão

## Aumentar o Dataset é a solução mais importante para combater o overffiting. O modelo precisa de muito mais exemplos de chaves em ângulos diferentes, com iluminações diferentes, fundos variados e tipos distintos;

## Podemos usar Data Augmentation e também, revisar a qualidade das anotações.

# Resumo do Teste Final - 50 Épocas:

### Box(P) - Precisão (0.564): Mede a exatidão das detecções. De todas as vezes que o modelo disse "isto é uma chave", ele estava certo em 56.4% das vezes. É um resultado moderado. Isso significa que para cada detecção correta, há quase uma detecção incorreta (falso positivo). O modelo ainda está um pouco "confuso".

### Box(R) - Recall (0.441): Mede a capacidade do modelo de encontrar todos os objetos existentes. De todas as chaves que realmente estavam nas imagens, o modelo conseguiu encontrar apenas 44.1%.Este resultado é um ponto crítico. Significa que o modelo deixou de detectar mais da metade das chaves que estavam presentes (falsos negativos). Em muitas aplicações, um recall baixo pode ser um problema maior que uma precisão baixa.

### mAP50 (0.534): É a métrica principal, que representa a nota geral do modelo, equilibrando precisão e recall. O "50" indica que uma detecção é considerada um acerto se a caixa prevista tiver pelo menos 50% de sobreposição com a caixa real. Uma pontuação de 53.4% classifica o modelo como um protótipo funcional. Ele aprendeu o conceito básico de "chave", mas ainda não é robusto o suficiente para ser considerado confiável, pois seu desempenho cai significativamente em dados novos.

### mAP50-95 (0.157): Uma métrica muito mais rigorosa que mede o desempenho em níveis de exigência de sobreposição de 50% até 95%.O valor baixo de 15.7% confirma que as poucas detecções corretas que o modelo fez geralmente não estavam bem ajustadas ao objeto. As caixas delimitadoras estavam um pouco "frouxas".

# Conclusão:

### O modelo "decorou" muito bem as imagens de treino, mas não aprendeu a generalizar esse conhecimento de forma eficaz para novas imagens. Ele funciona, mas ainda não é confiável. O resultado, no entanto, é um excelente diagnóstico que aponta o caminho para melhorias, como aumentar significativamente a variedade e a quantidade de imagens de treino.


# Configuração dos Caminhos e Preparação de Dados para o Dataset - XÍCARAS

In [4]:
# CÉLULA 2: Configuração dos Caminhos e Preparação do Dataset de XÍCARAS

import os
import json
import shutil
from tqdm import tqdm

# --- Definição dos novos caminhos para o dataset de xícaras ---
# Caminho base para a pasta 'xicaras'
gdrive_base_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras'

# Caminho para a pasta de imagens de TREINAMENTO
gdrive_image_path = os.path.join(gdrive_base_path, 'train')

# Nome do arquivo JSON em formato COCO para as xícaras
gdrive_json_filename = 'labels_my-project-name_2025-10-13-12-41-18.json'
gdrive_json_path = os.path.join(gdrive_base_path, gdrive_json_filename)

# --- Criação das pastas locais no Colab ---
base_dir = '/content/yolo_dataset_xicaras'
images_train_dir = os.path.join(base_dir, 'images', 'train')
labels_train_dir = os.path.join(base_dir, 'labels', 'train')

if os.path.exists(base_dir):
    shutil.rmtree(base_dir)

os.makedirs(images_train_dir, exist_ok=True)
os.makedirs(labels_train_dir, exist_ok=True)

print(f"Estrutura de pastas criada em: {base_dir}")

# --- Carregando o arquivo de anotações ---
with open(gdrive_json_path, 'r') as f:
    coco_data = json.load(f)

# --- Processamento das Anotações e Imagens ---
images_info = {img['id']: (img['file_name'], img['width'], img['height']) for img in coco_data['images']}
categories_info = {cat['id']: cat['name'] for cat in coco_data['categories']}
class_names = list(categories_info.values())
num_classes = len(class_names)

print(f"Dataset contém {num_classes} classe(s): {', '.join(class_names)}")

print("\nProcessando anotações e copiando imagens de xícaras...")
for ann in tqdm(coco_data['annotations']):
    image_id = ann['image_id']
    category_id = ann['category_id']
    bbox = ann['bbox']

    file_name, img_w, img_h = images_info[image_id]

    # Conversão do formato COCO para YOLO
    x_min, y_min, bbox_w, bbox_h = bbox
    x_center = (x_min + bbox_w / 2) / img_w
    y_center = (y_min + bbox_h / 2) / img_h
    norm_w = bbox_w / img_w
    norm_h = bbox_h / img_h
    yolo_class_id = category_id - 1
    yolo_format_str = f"{yolo_class_id} {x_center:.6f} {y_center:.6f} {norm_w:.6f} {norm_h:.6f}\n"

    # Salva o arquivo de anotação
    label_file_name = os.path.splitext(file_name)[0] + '.txt'
    with open(os.path.join(labels_train_dir, label_file_name), 'a') as f:
        f.write(yolo_format_str)

    # Copia o arquivo de imagem do Google Drive para a pasta local do Colab
    source_image_path = os.path.join(gdrive_image_path, file_name)
    dest_image_path = os.path.join(images_train_dir, file_name)
    if not os.path.exists(dest_image_path):
        shutil.copy(source_image_path, dest_image_path)

print("\nPreparação do dataset de xícaras concluída!")

Estrutura de pastas criada em: /content/yolo_dataset_xicaras
Dataset contém 1 classe(s): xicara

Processando anotações e copiando imagens de xícaras...


100%|██████████| 38/38 [00:00<00:00, 127.93it/s]


Preparação do dataset de xícaras concluída!


# Criando Arquivo de Configuração do Dataset (YAML)

In [5]:
import yaml

# Cria o dicionário com as informações do novo dataset
yaml_data = {
    'train': os.path.join(base_dir, 'images', 'train'),
    'val': os.path.join(base_dir, 'images', 'train'), # Usando o mesmo conjunto para validação
    'nc': num_classes,
    'names': class_names
}

# Salva o arquivo de configuração com um novo nome
yaml_file_path = '/content/dataset_xicaras.yaml'
with open(yaml_file_path, 'w') as f:
    yaml.dump(yaml_data, f, sort_keys=False)

print(f"Arquivo de configuração '{yaml_file_path}' criado com sucesso:")
!cat {yaml_file_path}

Arquivo de configuração '/content/dataset_xicaras.yaml' criado com sucesso:
train: /content/yolo_dataset_xicaras/images/train
val: /content/yolo_dataset_xicaras/images/train
nc: 1
names:
- xicara


# Iniciando o Treinamento de Modelo - XÍCARAS

In [6]:
from ultralytics import YOLO

# Carrega um modelo pré-treinado (yolov8n.pt é o menor e mais rápido)
model = YOLO('yolov8n.pt')

# Inicia o treinamento
results = model.train(
    data=yaml_file_path,       # Usa o novo arquivo .yaml que criamos
    epochs=50,
    imgsz=640,
    project='yolo_training',   # Pasta principal para os resultados
    name='xicaras_detector'    # Nome da subpasta específica para este treinamento
)

print("\nTreinamento de xícaras concluído!")
print("Os resultados, incluindo os pesos do modelo treinado, estão na pasta 'yolo_training/xicaras_detector'.")

Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
Ultralytics 8.3.214 🚀 Python-3.12.12 torch-2.8.0+cu126 CUDA:0 (Tesla T4, 15095MiB)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/dataset_xicaras.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=

# Resumo das métricas do Treinamento - 25 Épocas

### Box(P) - Precisão (0.971): Quando o modelo diz "isto é uma xícara", ele está correto em 97.1% das vezes. Um resultado fantástico, com pouquíssimos alarmes falsos.

### Box(R) - Recall (0.887): De todas as xícaras que realmente existiam nas imagens, o modelo conseguiu encontrar 88.7% delas. Também é um número muito alto, mostrando que o modelo raramente deixa passar uma xícara sem detectar.

### mAP50 (0.989): Esta é a métrica principal e o resultado é quase perfeito: 98.9%. Isso indica que o modelo é extremamente bom em localizar as xícaras corretamente. A performance é significativamente superior à do modelo de chaves na mesma etapa.

### mAP50-95 (0.906): Este é talvez o resultado mais impressionante. Um valor de 90.6% nesta métrica rigorosa é excepcional. Ele nos diz que as caixas delimitadoras que o modelo desenha não estão apenas no lugar certo, mas também estão muito bem ajustadas ao tamanho e formato das xícaras.

# Resumo das métricas do Treinamento - 50 Épocas:

### Box(P) - Precisão (0.999): Resultado Praticamente Perfeito. Uma precisão de 99.9% indica que quase todas as detecções que o modelo fez estavam corretas. Ele praticamente não cometeu erros de "falso positivo" (identificar um objeto que não era uma xícara). Isso o torna extremamente confiável.

### Box(R) - Recall (1): Resultado Perfeito. Um recall de 100% é a pontuação máxima. Significa que, de todas as xícaras que realmente existiam nas 38 imagens, o modelo encontrou TODAS. Ele não deixou nenhuma passar despercebida, o que é um feito fantástico.

### mAP50 (0.995): Desempenho de Elite. A métrica principal, que combina precisão e recall, atingiu 99.5%. Este é um resultado de altíssima qualidade, confirmando que o modelo localiza as xícaras com uma precisão quase perfeita.

### mAP50-95 (0.963): Excepcional e Robusto. ste é, talvez, o resultado mais impressionante. Um valor de 96.3% nesta métrica, que é muito mais rigorosa, é excepcional. Ele nos diz que as caixas delimitadoras que o modelo desenha não estão apenas no lugar certo, mas também estão muito bem ajustadas ao tamanho e formato das xícaras, mesmo sob critérios de avaliação muito exigentes.

# Conclusão:

### Este modelo de detecção de xícaras atingiu um nível de desempenho de elite. A combinação de precisão quase perfeita, recall perfeito e um mAP50-95 altíssimo indica que o modelo não apenas aprendeu a identificar xícaras, mas o fez de forma extremamente robusta e precisa.

### Diferente do modelo de chaves, que mostrou sinais de overfitting, este modelo parece ter aprendido as características essenciais das xícaras de uma maneira que o torna muito mais generalizável.

#Validação Dedicada - XÍCARAS

In [7]:
import os
import json
import shutil
import yaml
from ultralytics import YOLO
from tqdm import tqdm

print("--- Iniciando o Processo de Validação do Modelo de Xícaras ---")

# --- 1. CONFIGURAÇÃO DOS CAMINHOS ---
# Caminho para a pasta de IMAGENS de validação
gdrive_val_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras/validation'

# Caminho para a pasta base onde o JSON de validação está localizado
gdrive_base_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras'

# Nome do seu arquivo JSON de validação (formato COCO)
gdrive_val_json_name = 'labels_my-project-name_2025-10-14-04-17-44.json'
gdrive_val_json_path = os.path.join(gdrive_base_path, gdrive_val_json_name)

# Caminho para o seu melhor modelo TREINADO de xícaras
model_path = '/content/yolo_training/xicaras_detector/weights/best.pt'

# --- 2. PREPARAÇÃO DO DATASET DE VALIDAÇÃO ---
print(f"\n[PASSO 1/3] Preparando o dataset de validação a partir de: {gdrive_val_path}")

base_dir_val = '/content/yolo_dataset_xicaras_validation'
images_val_dir = os.path.join(base_dir_val, 'images', 'val')
labels_val_dir = os.path.join(base_dir_val, 'labels', 'val')

if os.path.exists(base_dir_val):
    shutil.rmtree(base_dir_val)
os.makedirs(images_val_dir, exist_ok=True)
os.makedirs(labels_val_dir, exist_ok=True)

try:
    print(f"Carregando anotações de: {gdrive_val_json_path}")
    with open(gdrive_val_json_path, 'r') as f:
        coco_data = json.load(f)

    if 'annotations' not in coco_data or 'images' not in coco_data or 'categories' not in coco_data:
        raise ValueError("Erro: O arquivo JSON não parece estar no formato COCO.")

    images_info = {img['id']: (img['file_name'], img['width'], img['height']) for img in coco_data['images']}
    categories_info = {cat['id']: cat['name'] for cat in coco_data['categories']}
    class_names = list(categories_info.values())
    num_classes = len(class_names)

    print(f"Encontradas {len(coco_data['images'])} imagens e {len(coco_data['annotations'])} anotações no JSON.")

    for ann in tqdm(coco_data['annotations'], desc="Processando anotações"):
        if 'category_id' not in ann or 'bbox' not in ann:
            continue

        image_id = ann['image_id']
        file_name, img_w, img_h = images_info[image_id]

        source_image_path = os.path.join(gdrive_val_path, file_name)
        dest_image_path = os.path.join(images_val_dir, file_name)
        if not os.path.exists(dest_image_path):
            shutil.copy(source_image_path, dest_image_path)

        label_file_name = os.path.splitext(file_name)[0] + '.txt'
        with open(os.path.join(labels_val_dir, label_file_name), 'a') as f:
            category_id = ann['category_id']
            bbox = ann['bbox']
            x_min, y_min, bbox_w, bbox_h = bbox
            x_center = (x_min + bbox_w / 2) / img_w
            y_center = (y_min + bbox_h / 2) / img_h
            norm_w = bbox_w / img_w
            norm_h = bbox_h / img_h
            yolo_class_id = category_id - 1
            yolo_format_str = f"{yolo_class_id} {x_center:.6f} {y_center:.6f} {norm_w:.6f} {norm_h:.6f}\n"
            f.write(yolo_format_str)

    print("Dataset de validação preparado com sucesso.")

    # --- 3. CRIAÇÃO DO ARQUIVO YAML ---
    print("\n[PASSO 2/3] Criando arquivo de configuração 'validation_xicaras.yaml'")
    yaml_val_data = {
        'train': images_val_dir,
        'val': images_val_dir,
        'nc': num_classes,
        'names': class_names
    }
    yaml_val_file_path = '/content/validation_xicaras.yaml'
    with open(yaml_val_file_path, 'w') as f:
        yaml.dump(yaml_val_data, f, sort_keys=False)
    !cat {yaml_val_file_path}

    # --- 4. EXECUÇÃO DA VALIDAÇÃO ---
    print("\n[PASSO 3/3] Executando a validação...")
    model = YOLO(model_path)
    validation_results = model.val(data=yaml_val_file_path)
    print("\n--- Processo de Validação Concluído ---")

except (FileNotFoundError, ValueError) as e:
    print(f"\nERRO CRÍTICO: {e}")

--- Iniciando o Processo de Validação do Modelo de Xícaras ---

[PASSO 1/3] Preparando o dataset de validação a partir de: /content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras/validation
Carregando anotações de: /content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras/labels_my-project-name_2025-10-14-04-17-44.json
Encontradas 5 imagens e 5 anotações no JSON.


Processando anotações: 100%|██████████| 5/5 [00:00<00:00, 313.79it/s]

Dataset de validação preparado com sucesso.

[PASSO 2/3] Criando arquivo de configuração 'validation_xicaras.yaml'
train: /content/yolo_dataset_xicaras_validation/images/val
val: /content/yolo_dataset_xicaras_validation/images/val
nc: 1
names:
- xicara

[PASSO 3/3] Executando a validação...
Ultralytics 8.3.214 🚀 Python-3.12.12 torch-2.8.0+cu126 CUDA:0 (Tesla T4, 15095MiB)


Model summary (fused): 72 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1585.7±389.1 MB/s, size: 61.8 KB)
val: Scanning /content/yolo_dataset_xicaras_validation/labels/val... 5 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 5/5 1.1Kit/s 0.0s
val: New cache created: /content/yolo_dataset_xicaras_validation/labels/val.cache
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 1/1 3.9it/s 0.3s
                   all          5          5       0.99          1      0.995      0.596
Speed: 0.3ms preprocess, 31.7ms inference, 0.0ms loss, 4.5ms postprocess per image
Results saved to /content/runs/detect/val

--- Processo de Validação Concluído ---


#Resumo de Validação Dedicada - XÍCARAS - 25 Épocas

### Box(P) - Precisão (0.949): A precisão de 94.9% é excelente. Significa que praticamente todas as detecções que o modelo fez foram corretas.

### Box(R) - Recall (1): Este é um resultado perfeito (100%). De todas as xícaras que existiam nas 5 imagens de validação, o modelo encontrou todas.

### mAP50 (0.995): A métrica principal é de 99.5%, o que é praticamente a pontuação máxima. Isso confirma que o modelo não só encontra todas as xícaras, mas também as localiza com altíssima precisão.

### mAP50-95 (0.551): Este resultado é o único ponto que merece atenção. Houve uma queda em relação ao treino (de 90.6% para 55.1%). O que significa que, embora o modelo encontre todas as xícaras, a "qualidade" das caixas delimitadoras (o quão justo o retângulo fica ao redor do objeto) não é tão perfeita nas imagens de validação quanto era nas de treino. No entanto, um valor de 55.1% ainda é considerado bom e funcional.

# Resumo de Validação Dedicada - XÍCARAS - 50 Épocas

### Box(P) - Precisão (0.99): Quase Perfeito. Mede a exatidão das detecções. De todas as vezes que o modelo disse "isto é uma xícara", ele estava certo em 99% das vezes. Este é um resultado de elite. Significa que o modelo é extremamente confiável e quase nunca comete erros de "falso positivo" (identificar um objeto que não é uma xícara).

### Box(R) - Recall (1): Perfeito. Mede a capacidade do modelo de encontrar todos os objetos existentes. Um recall de 100% é a pontuação máxima.O modelo encontrou TODAS as xícaras que existiam nas 5 imagens de validação. Ele não deixou nenhuma passar despercebida, o que é um feito fantástico e o torna muito robusto.

### mAP50 (0.995): Desempenho de Elite. A principal métrica, que combina precisão e recall. Uma detecção é considerada correta se a caixa prevista tiver pelo menos 50% de sobreposição com a caixa real. Uma pontuação de 99.5% é o mais próximo da perfeição que se pode esperar. Isso confirma que o modelo localiza as xícaras com uma precisão altíssima.

### mAP50-95 (0.596): Bom e Consistente. Uma métrica muito mais rigorosa, que calcula a média da performance em 10 níveis diferentes de exigência (de 50% a 95% de sobreposição). O valor de 59.6% é bom. A queda em relação ao valor do treinamento (que era de 96.3%) é esperada e normal. Isso nos diz que, embora o modelo encontre as xícaras perfeitamente, a delimitação das caixas delimitadoras não é sempre milimétrica em imagens novas, mas ainda assim é muito funcional.

# Conclusão:

### O modelo de detecção de xícaras é excelente e confiável. O fato de a métrica principal (mAP50) e o recall terem se mantido perfeitos na validação é a melhor prova de que ele não sofreu de overfitting. Ele aprendeu as características essenciais de uma "xícara" de forma tão robusta que consegue aplicar esse conhecimento com a mesma eficácia em imagens que nunca viu antes.

# Teste Modelo - XÍCARA

In [8]:
import os
import json
import shutil
import yaml
from ultralytics import YOLO
from tqdm import tqdm

print("--- Iniciando o Processo de Teste Final do Modelo de Xícaras ---")

# --- 1. CONFIGURAÇÃO DOS CAMINHOS ---
# Caminho para a pasta de IMAGENS de teste
gdrive_test_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras/test'

# Caminho para a pasta base onde o JSON de teste está localizado
gdrive_base_path = '/content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras'

# Nome do seu arquivo JSON de teste (formato COCO)
gdrive_test_json_name = 'labels_my-project-name_2025-10-14-04-32-33.json'
gdrive_test_json_path = os.path.join(gdrive_base_path, gdrive_test_json_name)

# Caminho para o seu melhor modelo TREINADO de xícaras
model_path = '/content/yolo_training/xicaras_detector/weights/best.pt'

# --- 2. PREPARAÇÃO DO DATASET DE TESTE ---
print(f"\n[PASSO 1/3] Preparando o dataset de teste a partir de: {gdrive_test_path}")

base_dir_test = '/content/yolo_dataset_xicaras_test'
images_test_dir = os.path.join(base_dir_test, 'images', 'test')
labels_test_dir = os.path.join(base_dir_test, 'labels', 'test')

if os.path.exists(base_dir_test):
    shutil.rmtree(base_dir_test)
os.makedirs(images_test_dir, exist_ok=True)
os.makedirs(labels_test_dir, exist_ok=True)

try:
    print(f"Carregando anotações de: {gdrive_test_json_path}")
    with open(gdrive_test_json_path, 'r') as f:
        coco_data = json.load(f)

    if 'annotations' not in coco_data or 'images' not in coco_data or 'categories' not in coco_data:
        raise ValueError("Erro: O arquivo JSON não parece estar no formato COCO.")

    images_info = {img['id']: (img['file_name'], img['width'], img['height']) for img in coco_data['images']}
    categories_info = {cat['id']: cat['name'] for cat in coco_data['categories']}
    class_names = list(categories_info.values())
    num_classes = len(class_names)

    print(f"Encontradas {len(coco_data['images'])} imagens e {len(coco_data['annotations'])} anotações no JSON de teste.")

    for ann in tqdm(coco_data['annotations'], desc="Processando anotações de teste"):
        if 'category_id' not in ann or 'bbox' not in ann:
            continue

        image_id = ann['image_id']
        file_name, img_w, img_h = images_info[image_id]

        source_image_path = os.path.join(gdrive_test_path, file_name)
        dest_image_path = os.path.join(images_test_dir, file_name)
        if not os.path.exists(dest_image_path):
            shutil.copy(source_image_path, dest_image_path)

        label_file_name = os.path.splitext(file_name)[0] + '.txt'
        with open(os.path.join(labels_test_dir, label_file_name), 'a') as f:
            category_id = ann['category_id']
            bbox = ann['bbox']
            x_min, y_min, bbox_w, bbox_h = bbox
            x_center = (x_min + bbox_w / 2) / img_w
            y_center = (y_min + bbox_h / 2) / img_h
            norm_w = bbox_w / img_w
            norm_h = bbox_h / img_h
            yolo_class_id = category_id - 1
            yolo_format_str = f"{yolo_class_id} {x_center:.6f} {y_center:.6f} {norm_w:.6f} {norm_h:.6f}\n"
            f.write(yolo_format_str)

    print("Dataset de teste preparado com sucesso.")

    # --- 3. CRIAÇÃO DO ARQUIVO YAML DE TESTE ---
    print("\n[PASSO 2/3] Criando arquivo de configuração 'test_xicaras.yaml'")
    yaml_test_data = {
        'train': images_test_dir,
        'val': images_test_dir,   # Apontamos 'val' para o nosso conjunto de teste
        'nc': num_classes,
        'names': class_names
    }
    yaml_test_file_path = '/content/test_xicaras.yaml'
    with open(yaml_test_file_path, 'w') as f:
        yaml.dump(yaml_test_data, f, sort_keys=False)
    !cat {yaml_test_file_path}

    # --- 4. EXECUÇÃO DO TESTE FINAL ---
    print("\n[PASSO 3/3] Executando o teste final com o modelo treinado...")
    model = YOLO(model_path)
    test_results = model.val(data=yaml_test_file_path)
    print("\n--- Processo de Teste Concluído ---")

except (FileNotFoundError, ValueError) as e:
    print(f"\nERRO CRÍTICO: {e}")

--- Iniciando o Processo de Teste Final do Modelo de Xícaras ---

[PASSO 1/3] Preparando o dataset de teste a partir de: /content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras/test
Carregando anotações de: /content/drive/MyDrive/computer_vision_cap1/Computer_Vision_Cap1/xicaras/labels_my-project-name_2025-10-14-04-32-33.json
Encontradas 5 imagens e 5 anotações no JSON de teste.


Processando anotações de teste: 100%|██████████| 5/5 [00:00<00:00, 318.62it/s]

Dataset de teste preparado com sucesso.

[PASSO 2/3] Criando arquivo de configuração 'test_xicaras.yaml'
train: /content/yolo_dataset_xicaras_test/images/test
val: /content/yolo_dataset_xicaras_test/images/test
nc: 1
names:
- xicara

[PASSO 3/3] Executando o teste final com o modelo treinado...
Ultralytics 8.3.214 🚀 Python-3.12.12 torch-2.8.0+cu126 CUDA:0 (Tesla T4, 15095MiB)


Model summary (fused): 72 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1926.6±413.2 MB/s, size: 62.6 KB)
val: Scanning /content/yolo_dataset_xicaras_test/labels/test... 5 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 5/5 1.2Kit/s 0.0s
val: New cache created: /content/yolo_dataset_xicaras_test/labels/test.cache
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 1/1 6.1it/s 0.2s
                   all          5          5      0.991          1      0.995      0.671
Speed: 0.6ms preprocess, 6.5ms inference, 0.0ms loss, 3.2ms postprocess per image
Results saved to /content/runs/detect/val2

--- Processo de Teste Concluído ---


#Resumo do Teste Final do Modelo - XÍCARAS - 25 Épocas

### Box(P) - Precisão (1): Uma precisão de 100% significa que todas as detecções quemodelo fez estavam corretas. Ele não cometeu nenhum erro de "falso positivo" (ou seja, nunca identificou um objeto que não era uma xícara). Isso torna o modelo extremamente confiável.

### Box(R) - Recall (0.976): Um resultado quase perfeito (97.6%). O modelo encontrou praticamente todas as xícaras que existiam nas imagens de teste. É um desempenho extremamente robusto.

### mAP50 (0.995): A métrica geral de desempenho é de 99.5%, o que é o mais próximo da perfeição que se pode esperar. Isso confirma que o modelo é extremamente preciso na localização das xícaras.

### mAP50-95 (0.55): Este resultado permaneceu consistente com o da validação (55%). Isso mostra que, embora o modelo encontre os objetos com precisão quase perfeita, a dimensão das caixas delimitadoras é boa, mas não perfeita, o que é um comportamento totalmente normal e aceitável para um modelo de alto desempenho.

# Conclusão

### O fato de as métricas de teste serem quase idênticas (e em precisão, até melhores) às da validação é a prova definitiva de que o modelo generalizou de forma brilhante. Ele não apenas "decorou" as imagens de treino, mas realmente aprendeu as características que definem uma xícara e consegue aplicar esse conhecimento de forma confiável a imagens completamente novas.

# Resumo do Teste Final do Modelo - XÍCARAS - 50 Épocas

### Box(P) - Precisão (0.991): Quase Perfeito.Mede a exatidão das detecções. De todas as vezes que o modelo disse "isto é uma xícara", ele estava certo em 99.1% das vezes. Este é um resultado de elite. Significa que o modelo é extremamente confiável e quase nunca comete erros de "falso positivo" (identificar um objeto que não é uma xícara).

### Mede a capacidade do modelo de encontrar todos os objetos existentes. Um recall de 100% é a pontuação máxima. O modelo encontrou todas as xícaras que existiam nas 5 imagens de teste. Ele não deixou nenhuma passar despercebida, o que é um feito fantástico e o torna muito robusto.


### mAP50 (0.995): Desempenho de Elite. A sua principal métrica, que combina precisão e recall. Uma detecção é considerada correta se a caixa prevista tiver pelo menos 50% de sobreposição com a caixa real.Uma pontuação de 99.5% é o mais próximo da perfeição que se pode esperar. Isso confirma que o modelo localiza as xícaras com uma precisão altíssima.


### mAP50-95 (0.671): Muito Bom e Consistente. Uma métrica muito mais rigorosa, que calcula a média da performance em 10 níveis diferentes de exigência (de 50% a 95% de sobreposição). O valor de 67.1% é muito bom e, mais importante, é consistente e até melhor que o resultado da validação (que foi 59.6%). Isso mostra que as caixas delimitadoras são bem ajustadas e que o modelo manteve sua qualidade sob critérios mais rigorosos.


# Conclusão:

### O resultado mais importante aqui é a consistência e a robustez. O fato de as métricas de teste serem praticamente idênticas (ou até melhores) às da validação é a prova definitiva de que o modelo não sofreu de overfitting e generaliza de forma brilhante. Ele não apenas "decorou" as imagens de treino; ele realmente "aprendeu" o que é uma xícara e consegue aplicar esse conhecimento com a mesma eficácia em imagens completamente novas.

# Conclusão Definitiva - Comparativo Chaves x Xícaras


## Ao comparar os dois modelos, a conclusão mais importante não é apenas que um é "melhor" que o outro, mas por que isso aconteceu:

### A qualidade e a variedade dos dados de treinamento são o fator mais determinante para o sucesso de um modelo de visão computacional.

### O trabalho demonstrou isso perfeitamente na prática. O modelo de xícaras foi um sucesso porque os dados permitiram que ele generalizasse. O modelo de chaves, apesar de usar a mesma tecnologia e o mesmo processo, teve dificuldades porque seus dados de treino não foram suficientes para prepará-lo para a variedade do mundo real.

### Houve sucesso não apenas o treinamento de dois modelos, mas também uma jornada completa de diagnóstico e análise.